internal-v2.nix

```diff
@@ -0,0 +1,556 @@
+        inherit parts;
+        org = builtins.elemAt parts 2;
+        repo = builtins.elemAt parts 4;
+        rev = builtins.elemAt parts 6;
+      };
+      partLen = builtins.length parts;
+    in
+    assert !((partLen == 13) || (partLen == 7)) ->
+      throw "failed to parse GitHub reference `${str}`. Expected a string of format `git+ssh://git@github.org/owner/repo.git#revision or `github:owner/repo#revision`";
+    if (partLen == 13)
+    then ghGitSshFormatParser str
+    else
+      if (builtins.elemAt parts 0) == "github"
+      then ghShortnameParser str
+      else ghGitRefParser str;
+
+  # Description: Takes an attribute set describing a git dependency and returns
+  # a .tgz of the repository as store path. If the attribute hash contains a
+  # hash attribute it will provide the value to `fetchFromGitHub` which will
+  # also work in restricted evaluation.
+  # Type: { name :: String , org :: String, repo :: String, rev :: String, ref :: String, hash :: String, sourceOptions :: Set }
+  #       -> drv
+  buildTgzFromGitHub = { name, org, repo, rev, ref, hash ? null, sourceOptions ? { } }:
+    let
+      src =
+        if hash != null then
+          fetchFromGitHub
+            {
+              owner = org;
+              inherit repo;
+              inherit rev;
+              sha256 = hash; # FIXME: what if sha3?
+            } else
+          fetchGitWrapped {
+            url = "https://github.com/${org}/${repo}";
+            inherit rev ref;
+            allRefs = true;
+          };
+
+      sourceInfo = {
+        github = { inherit org repo rev ref; };
+      };
+      drv = packTgz sourceOptions.nodejs name ref src;
+    in
+    if sourceOptions ? sourceOverrides.${name}
+    then sourceOptions.sourceOverrides.${name} sourceInfo drv
+    else drv;
+
+  # Description: Packs a source directory into a .tgz tar archive. If the
+  # source is an archive, it gets unpacked first.
+  # Type: Path -> String -> String -> Path -> Path
+  packTgz = nodejs: pname: version: src: stdenv.mkDerivation (
+    let
+      preInstallLinks = writeShellScript "preInstallLinks" ''
+        # preinstalled.links is a space separated text file in the
+        # form of:
+        # $symlink-target $symlink-location
+        for link in "$(<preinstalled.links)"; do
+          if [ ! -z "$link" ]; then
+            target="$(echo "$link" | cut -d ' ' -f 1)"
+            location="$(echo "$link" | cut -d ' ' -f 2)"
+            mkdir -p "$(dirname "$location")"
+            ln -sf "$target" "$location"
+          fi
+        done
+      '';
+    in
+    rec {
+      name = lib.strings.sanitizeDerivationName "${pname}-${version}.tgz";
+      id = "id_" + (builtins.replaceStrings [ "-" "." ] [ "_" "_" ] name);
+
+      phases = "unpackPhase patchPhase installPhase";
+
+      inherit src;
+      sourceRoot = "package";
+      outputs = [ "out" "hash" ];
+      nativeBuildInputs = [ jq openssl ];
+
+      propagatedBuildInputs = [
+        nodejs
```

Comment:
I used coreutil to patch binaries to avoid having to rebuild packages when changing nodejs version. It worked but might not be the right solution for this.